### ISSUE TITLE
imtest with opencv4.0.1

### ISSUE THREAD

**Issue by tomsang** [pjreddie/darknet#1421]
Following the patch https://github.com/pjreddie/darknet/commit/24cff08086b573c3341e91d072430c9f624a2208#diff-25d902c24283ab8cfbac54dfa101ad31,
I successfully built darknet.  But I found 2 bad images(Original and c1) when I run ./darknet imtest data/eagle.jpg.
See the attached picture.
![badimtest](https://user-images.githubusercontent.com/15348764/52535716-cb539780-2d1f-11e9-80f6-515b724c8457.jpg)

**Comment by nkalupahana** [comment]
Does it do this with other images as well?

**Comment by tomsang** [comment]
@nkalupahana 
The issue happens on all images under data\
@tiagoshibata Did you meet the same issue with your patch ?

**Comment by tiagoshibata** [comment]
No, I didn't. I will test it more extensively and try `imtest` and report my results.

**Comment by tiagoshibata** [comment]
Works perfectly for me, using the mainstream OpenCV 4.0.1-5 for Arch Linux.

**Comment by tiagoshibata** [comment]
Ok, I can reproduce it. Bug seems unrelated to my fix - if you open `image.c` and save the images using `save_image`, you get the correct output. Only the windowed display is corrupted.

A workaround is to redraw all windows at every update in `image.c`. Move these lines: [image.c#L1261-L1266](https://github.com/pjreddie/darknet/blob/24cff08086b573c3341e91d072430c9f624a2208/src/image.c#L1261-L1266) inside the `while(1){` loop at line 1268.

**Comment by tomsang** [comment]
I tried Update 2.  It works well!  
Thanks!

**Comment by tomsang** [comment]
It's better to find  why OpenCV4.0.1 has this issue.

**Comment by luoan** [comment]
> Ok, I can reproduce it. Bug seems unrelated to my fix - if you open `image.c` and save the images using `save_image`, you get the correct output. Only the windowed display is corrupted.
> 
> A workaround is to redraw all windows at every update in `image.c`. Move these lines: [image.c#L1261-L1266](https://github.com/pjreddie/darknet/blob/24cff08086b573c3341e91d072430c9f624a2208/src/image.c#L1261-L1266) inside the `while(1){` loop at line 1268.
> 
> Update: Garbled output happens both under X and under Wayland, unless the window is continuously updated.
> 
> Update 2: Seems like OpenCV is corrupting the first call to `imshow`. On my machine, adding a dummy call to `imshow` by duplicating line [image.c#L1261](https://github.com/pjreddie/darknet/blob/24cff08086b573c3341e91d072430c9f624a2208/src/image.c#L1261) also works around the issue.


hi, why did you use free in src/image_opencv.cpp line 30? 

`12 Mat image_to_mat(image im)
 13 {
 14     assert(im.c == 3 || im.c == 1);
 15     int x,y,c;
 16     image copy = copy_image(im);
 17     constrain_image(copy);
 18     if(im.c == 3) rgbgr_image(copy);
 19     unsigned char *data = (unsigned char *)malloc(im.w * im.h * im.c);
 20     for(y = 0; y < im.h; ++y){
 21         for(x = 0; x < im.w; ++x){
 22             for(c= 0; c < im.c; ++c){
 23                 float val = copy.data[c*im.h*im.w + y*im.w + x];
 24                 data[y*im.w*im.c + x*im.c + c] = (unsigned char)(val*255);
 25             }
 26         }
 27     }
 28     Mat m(im.h, im.w, CV_MAKETYPE(CV_8U, im.c), data);
 29     free_image(copy);
 30     //free(data);
 31     return m;
 32 }`

if commented out free(data) in line 30 on src/image_opencv.cpp
 ./darknet imtest data/eagle.jpg works fine for me. 
 i think opencv just use your original pointer, and did not do any copy. so if use free, system will reclaims memory of the picutre until some value of rgb value is 0, it regard it as \0 and end reclaims memory, so some value on the top of the image is meaningless. 

[https://docs.opencv.org/4.0.1/d3/d63/classcv_1_1Mat.html](url)
Mat() [11/29]   
`
data | Pointer to the user data. Matrix constructors that take data and step parameters do not allocate matrix data. Instead, they just initialize the matrix header that points to the specified data, which means that no data is copied. This operation is very efficient and can be used to process external data using OpenCV functions. The external data is not automatically deallocated, so you should take care of it.
`
this is my idea

**Comment by luoan** [comment]
> It's better to find why OpenCV4.0.1 has this issue.
does it helpful if you 
add
m = m.clonde();
after
Mat m(im.h, im.w, CV_MAKETYPE(CV_8U, im.c), data);
in src/image_opencv.cpp
and rebuild darknet

**Comment by luoan** [comment]
> It's better to find why OpenCV4.0.1 has this issue.

you can try endit    src/image_opencv.cpp    add a line
m = m.clonde();
after
Mat m(im.h, im.w, CV_MAKETYPE(CV_8U, im.c), data);
this might be helpful

**Comment by luoan** [comment]
> Ok, I can reproduce it. Bug seems unrelated to my fix - if you open `image.c` and save the images using `save_image`, you get the correct output. Only the windowed display is corrupted.
> 
> A workaround is to redraw all windows at every update in `image.c`. Move these lines: [image.c#L1261-L1266](https://github.com/pjreddie/darknet/blob/24cff08086b573c3341e91d072430c9f624a2208/src/image.c#L1261-L1266) inside the `while(1){` loop at line 1268.
> 
> Update: Garbled output happens both under X and under Wayland, unless the window is continuously updated.
> 
> Update 2: Seems like OpenCV is corrupting the first call to `imshow`. On my machine, adding a dummy call to `imshow` by duplicating line [image.c#L1261](https://github.com/pjreddie/darknet/blob/24cff08086b573c3341e91d072430c9f624a2208/src/image.c#L1261) also works around the issue.

i think if you want to save image after conver from struct image to class cv::Mat, we need to do it in  show_image_cv function in src/image_opencv.cpp.   
 not in src/image.cpp, in src/image.cpp call save_image function, save_image function will not call image_to_mat

**Comment by tiagoshibata** [comment]
@luoan good catch :smile: , I pushed a fix at https://github.com/tiagoshibata/darknet/commit/795e5ac1ebee56dac1f792938a9698604e9e4638

**Comment by luoan** [comment]
> @luoan good catch  , I pushed a fix at [tiagoshibata@795e5ac](https://github.com/tiagoshibata/darknet/commit/795e5ac1ebee56dac1f792938a9698604e9e4638)

great solution, your solution is very efficient !
